design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+
+After installing KSQL, developers are met with an immediate hurdle - they must have existing data 
+in Kafka that match the data they expect to handle. In a relational database environment, developers
+can use `INSERT INTO ... VALUES` to quickly construct a toy environment for testing. Generating this 
+data KSQL, however, is cumbersome: 
+
+* you require a separate tool (e.g. `kafkacat`, `ksql-datagen`)
+* you must know the deployment topology (e.g. URIs for ZK/Broker)
+* there is no synergy with KSQL (e.g. tools are not *schema aware* and can produce arbitrary data)
+
+This KLIP attempts to integrate producing test data into the KSQL REPL.
+
+## Scope
+
+This KLIP is intended to improve the *development and testing* workflow as opposed to providing
+*production-ready* functionality. The following tasks will be addressed:
+
+* KSQL will be able to create and produce data to arbitrary topics
+* KSQL will be able to validate data when producing to a topic that is registered with a schema
+in KSQL (e.g. as a result of `CREATE STREAM`)
+
+In the future, this functionality can be extended to be "production grade", allowing system
+operators or first-responders to quickly patch up data holes in case of emergency. This KLIP will
+explicitly discourage such usage in documentation (as it needs further testing before we can
+recommend it), but will not artificially impose a restriction on such usage.
+
+## Value
+
+This proposal will improve the first 15-minute experience that developers have with KSQL and will
+help seasoned KSQL developers test new functionality. Furthermore, this functionality can be
+leveraged to create a powerful testing harness for KSQL applications that can run in a sandbox
+without pre-existing data.
+
+## Public APIS
+
+#### INSERT INTO ... VALUES
+
+```sql
+INSERT INTO <stream_name|table_name> [(column1, column2, ...)]
+  VALUES (value1, value2, ...)
+```
+
+The `INSERT VALUES` statement can be used to insert new records into a KSQL source (agnostic to the
+Stream/Table duality). If the columns are not present, it is assumed that the values are in the same
+order as the schema and contain every field. `ROWKEY` is always present as the first column, and is 
+automatically populated with semantics described in the section on `ROWKEY` and `ROWTIME` below.
+
+The value for `stream_name`/`table_name` must be a valid Stream/Table registered with a KSQL schema.
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (see _Complex Datatypes_
+section below).
+
+Future extensions: 
+* Support Structs/Complex Types
+* Support some `AT INTERVAL` syntax to generalize the functionality to generate full stream data.
+* Read from file to allow reproducible insertions
+* Headless mode support
```

Comment:
Baby steps - there was contention about whether or not to introduce it and there is a clear path from one state to another, so for the first implementation it will not be supported in headless mode (also with my implementation, it is more work to support it as opposed to more work to restrict it)

Reply:
Why is adding this to headless contentious?  I'd be interested to hear what people are concerned about...

Reply:
If you have more than one server or restart your server it would start sending duplicate events - that's kind of weird semantics.

Reply:
I don't think that's a valid reason to not include this as the fix is super simple: one have `INSERT VALUES` statements in one of the servers sql file. It's an obvious thing for users IMHO.

Users can even have different headless nodes responsible for inserting different data sets if they want.  And yes, if they have multiple servers with the same insert statements... they'll get inserted multiple times... obviously!  Documentation is all that's needed here.

Reply:
> I don't think that's a valid reason to not include this as the fix is super simple: one have INSERT VALUES statements in one of the servers sql file. It's an obvious thing for users IMHO.

I thought about that, but it doesn't help the server restart issue. Plus if you're running this on some k8 cluster or another cloud environment that automatically scales/rebalances/picks up servers moves things around every time it gets moved you will insert the duplicate values. 

Let me phrase the question another way to understand your viewpoint - why is it bad to limit this functionality to interactive mode only for now?